File: android/src/main/java/com/cantoo/capacitor/onnx/SessionManager.kt
package com.cantoo.capacitor.onnx

import ai.onnxruntime.OrtEnvironment
import ai.onnxruntime.OrtSession
import java.util.concurrent.ConcurrentHashMap

data class SessionConfig(
    val executionProvider: String = "auto",
    val intraOpNumThreads: Int? = null,
    val interOpNumThreads: Int? = 1,
) {
    fun normalized(): SessionConfig {
        val provider = when (executionProvider.lowercase()) {
            "cpu", "nnapi", "auto" -> executionProvider.lowercase()
            else -> "auto"
        }
        val intra = intraOpNumThreads?.takeIf { it > 0 }
        val inter = interOpNumThreads?.takeIf { it > 0 }
        return SessionConfig(provider, intra, inter)
    }
}

data class SessionRef(
    val modelId: String,
    val version: String,
    val session: OrtSession,
    val config: SessionConfig,
    val executionProviderUsed: String,
)

class SessionManager {
    private val environment: OrtEnvironment = OrtEnvironment.getEnvironment()
    private val sessions = ConcurrentHashMap<String, SessionRef>()

    fun ortEnvironment(): OrtEnvironment = environment

    fun ensureSession(modelId: String, version: String, filePath: String, config: SessionConfig): SessionRef {
        val key = key(modelId, version)
        val normalized = config.normalized()

        synchronized(this) {
            val existing = sessions[key]
            if (existing != null && existing.config == normalized) {
                return existing
            }

            if (existing != null) {
                existing.session.close()
                sessions.remove(key)
            }

            val created = createSession(modelId, version, filePath, normalized)
            sessions[key] = created
            return created
        }
    }

    fun getSession(modelId: String, version: String): SessionRef? = sessions[key(modelId, version)]

    fun closeSession(modelId: String, version: String) {
        sessions.remove(key(modelId, version))?.session?.close()
    }

    fun closeAll() {
        sessions.values.forEach { it.session.close() }
        sessions.clear()
    }

    private fun createSession(modelId: String, version: String, filePath: String, config: SessionConfig): SessionRef {
        return when (config.executionProvider) {
            "cpu" -> createCpuSession(modelId, version, filePath, config)
            "nnapi" -> createNnapiSession(modelId, version, filePath, config)
            else -> createAutoSession(modelId, version, filePath, config)
        }
    }

    private fun createAutoSession(modelId: String, version: String, filePath: String, config: SessionConfig): SessionRef {
        return try {
            createNnapiSession(modelId, version, filePath, config)
        } catch (_: Throwable) {
            createCpuSession(modelId, version, filePath, config)
        }
    }

    private fun createNnapiSession(modelId: String, version: String, filePath: String, config: SessionConfig): SessionRef {
        val options = baseOptions(config)
        if (!tryEnableNnapi(options)) {
            throw IllegalStateException("SESSION_INIT_ERROR: NNAPI provider is not available on this device/build")
        }
        val session = environment.createSession(filePath, options)
        return SessionRef(modelId, version, session, config, "nnapi")
    }

    private fun createCpuSession(modelId: String, version: String, filePath: String, config: SessionConfig): SessionRef {
        val options = baseOptions(config)
        val session = environment.createSession(filePath, options)
        return SessionRef(modelId, version, session, config, "cpu")
    }

    private fun baseOptions(config: SessionConfig): OrtSession.SessionOptions {
        val defaultIntra = Runtime.getRuntime().availableProcessors().coerceAtMost(4).coerceAtLeast(1)
        val intra = config.intraOpNumThreads ?: defaultIntra
        val inter = config.interOpNumThreads ?: 1

        return OrtSession.SessionOptions().apply {
            setOptimizationLevel(OrtSession.SessionOptions.OptLevel.ALL_OPT)
            setIntraOpNumThreads(intra)
            setInterOpNumThreads(inter)
        }
    }

    private fun tryEnableNnapi(options: OrtSession.SessionOptions): Boolean {
        return try {
            val addNnapi = options::class.java.methods.firstOrNull { method ->
                method.name == "addNnapi" && method.parameterCount == 0
            } ?: return false
            addNnapi.invoke(options)
            true
        } catch (_: Throwable) {
            false
        }
    }

    private fun key(modelId: String, version: String): String = "$modelId::$version"
}

File: ios/Plugin/SessionManager.swift
import Foundation
import onnxruntime

struct SessionConfig {
    let executionProvider: String
    let intraOpNumThreads: Int?
    let interOpNumThreads: Int?

    init(executionProvider: String = "auto", intraOpNumThreads: Int? = nil, interOpNumThreads: Int? = nil) {
        self.executionProvider = executionProvider.lowercased()
        self.intraOpNumThreads = intraOpNumThreads
        self.interOpNumThreads = interOpNumThreads
    }

    func normalized() -> SessionConfig {
        let provider: String
        switch executionProvider {
        case "cpu", "coreml", "nnapi", "auto": provider = executionProvider
        default: provider = "auto"
        }
        return SessionConfig(
            executionProvider: provider,
            intraOpNumThreads: intraOpNumThreads.flatMap { $0 > 0 ? $0 : nil },
            interOpNumThreads: interOpNumThreads.flatMap { $0 > 0 ? $0 : nil }
        )
    }
}

struct SessionRef {
    let modelId: String
    let version: String
    let session: ORTSession
    let config: SessionConfig
    let executionProviderUsed: String
}

class SessionManager {
    private let environment: ORTEnvironment
    private var sessions: [String: SessionRef] = [:]
    private let lock = NSLock()

    init() throws {
        self.environment = try ORTEnvironment(loggingLevel: .warning)
    }

    func ensureSession(modelId: String, version: String, filePath: String, config: SessionConfig) throws -> SessionRef {
        let key = sessionKey(modelId, version)
        let normalized = config.normalized()

        lock.lock()
        defer { lock.unlock() }

        if let existing = sessions[key], configsMatch(existing.config, normalized) {
            return existing
        }

        sessions.removeValue(forKey: key)

        let created = try createSession(modelId: modelId, version: version, filePath: filePath, config: normalized)
        sessions[key] = created
        return created
    }

    func getSession(modelId: String, version: String) -> SessionRef? {
        lock.lock()
        defer { lock.unlock() }
        return sessions[sessionKey(modelId, version)]
    }

    func closeSession(modelId: String, version: String) {
        lock.lock()
        defer { lock.unlock() }
        sessions.removeValue(forKey: sessionKey(modelId, version))
    }

    func closeAll() {
        lock.lock()
        defer { lock.unlock() }
        sessions.removeAll()
    }

    private func createSession(modelId: String, version: String, filePath: String, config: SessionConfig) throws -> SessionRef {
        switch config.executionProvider {
        case "cpu":
            return try createCpuSession(modelId: modelId, version: version, filePath: filePath, config: config)
        case "coreml", "nnapi":
            return try createCoreMLSession(modelId: modelId, version: version, filePath: filePath, config: config)
        default:
            return try createAutoSession(modelId: modelId, version: version, filePath: filePath, config: config)
        }
    }

    private func createAutoSession(modelId: String, version: String, filePath: String, config: SessionConfig) throws -> SessionRef {
        do {
            return try createCoreMLSession(modelId: modelId, version: version, filePath: filePath, config: config)
        } catch {
            return try createCpuSession(modelId: modelId, version: version, filePath: filePath, config: config)
        }
    }

    private func createCoreMLSession(modelId: String, version: String, filePath: String, config: SessionConfig) throws -> SessionRef {
        let options = try buildOptions(config: config)
        do {
            try options.appendCoreMLExecutionProvider(withFlags: 0)
        } catch {
            throw OnnxPluginError.sessionInitError("CoreML provider is not available on this device")
        }
        let session = try ORTSession(env: environment, modelPath: filePath, sessionOptions: options)
        return SessionRef(modelId: modelId, version: version, session: session, config: config, executionProviderUsed: "coreml")
    }

    private func createCpuSession(modelId: String, version: String, filePath: String, config: SessionConfig) throws -> SessionRef {
        let options = try buildOptions(config: config)
        let session = try ORTSession(env: environment, modelPath: filePath, sessionOptions: options)
        return SessionRef(modelId: modelId, version: version, session: session, config: config, executionProviderUsed: "cpu")
    }

    private func buildOptions(config: SessionConfig) throws -> ORTSessionOptions {
        let options = try ORTSessionOptions()
        try options.setGraphOptimizationLevel(.all)
        let defaultIntra = min(ProcessInfo.processInfo.activeProcessorCount, 4)
        let intra = config.intraOpNumThreads ?? defaultIntra
        let inter = config.interOpNumThreads ?? 1
        try options.setIntraOpNumThreads(Int32(intra))
        try options.setInterOpNumThreads(Int32(inter))
        return options
    }

    private func configsMatch(_ a: SessionConfig, _ b: SessionConfig) -> Bool {
        return a.executionProvider == b.executionProvider &&
            a.intraOpNumThreads == b.intraOpNumThreads &&
            a.interOpNumThreads == b.interOpNumThreads
    }

    private func sessionKey(_ modelId: String, _ version: String) -> String {
        return "\(modelId)::\(version)"
    }
}
